app/src/main/java/com/example/travelapp/BinaryMinHeap.java

```diff
@@ -0,0 +1,166 @@
+
+    }
+
+    /*
+        placeElementAtIndex is a helper method that places the element in the passed in index where
+        it should belong in the heap. called by decreaseKey and add.
+     */
+
+    private int placeElementAtIndex(int index, int parent) {
+        if (arrayList.size() == 1) {
+            return 0;
+        }
+        while (index > 0 && arrayList.get(parent).key.compareTo(arrayList.get(index).key) > 0) {
+            //exchange the entry at parent and entry at index in arrayList and in indices
+            Entry<Key, V> parentKey = arrayList.get(parent);
+            Entry<Key, V> indexKey = arrayList.get(index);
+            arrayList.set(parent, indexKey);
+            arrayList.set(index, parentKey);
+            indices.replace(indexKey.value, parent);
+            indices.replace(parentKey.value, index);
+            // updates the index value
+            index = parent;
+            // calculates the new index of the parent of the current node,
+            // given the current node's index
+            Function<Integer, Integer> calculateParentIndex = (n) -> { return (n + 1) / 2 - 1; };
+            parent = calculateParentIndex.apply(index);
+        }
+        return index;
+    }
+
+    public void decreaseKey(V value, Key newKey) {
+        if (!this.containsValue(value)) {
+            throw new NoSuchElementException("value is not in the heap");
+        }
+        int index = indices.get(value);
+        if (index >= arrayList.size()) {
+            throw new IllegalArgumentException();
+        }
+        Key currKey = arrayList.get(index).key;
+        if (newKey == null) {
+            throw new IllegalArgumentException("newKey is null");
+        }
+        if (newKey.compareTo(currKey) > 0) {
+            throw new IllegalArgumentException("newKey > key(value)");
+        }
+
+        // update arraylist
+        arrayList.set(index, new Entry<Key, V>(newKey, value));
+        // calculates the new index of the parent of the current node,
+        // given the current node's index
+        Function<Integer, Integer> calculateParentIndex = (n) -> { return (n + 1) / 2 - 1; };
+        int parent = calculateParentIndex.apply(index);
+        // call helper function to place newKey in the right place
+        int finIndex = placeElementAtIndex(index, parent);
+        // update indices hashmap
+        indices.remove(value);
+        indices.put(value, finIndex);
+    }
+
+    public Entry<Key, V> extractMin() {
+        if (arrayList.size() == 0) {
+            throw new NoSuchElementException("the min-heap is empty");
+        }
+        Entry<Key, V> min = arrayList.get(0);
+        int len = arrayList.size();
+        Entry<Key, V> lastEntry = arrayList.get(len - 1);
+        // swap the minimum Entry and last Entry
+        arrayList.set(0, lastEntry);
+        arrayList.set(len - 1, min);
+        // remove minimum element
+        arrayList.remove(len - 1);
```

Comment:
curious, so if you wanted the minimum value the whole time, why not do a max heap so you don't have to go through the trouble of swapping the first/last and re "heapifying" the array?